src/include/souffle/datastructure/EquivalenceRelation.h

```diff
@@ -128,17 +128,28 @@ class EquivalenceRelation {
     }
 
     /**
-     * Extend this relation with another relation, expanding this equivalence relation
-     * The supplied relation is the old knowledge, whilst this relation only contains
-     * explicitly new knowledge. After this operation the "implicitly new tuples" are now
-     * explicitly inserted this relation.
+     * Extend this relation with another relation, expanding this equivalence
+     * relation and inserting it into the other relation.
+     *
+     * The supplied relation is the old knowledge, whilst this relation only
+     * contains explicitly new knowledge. After this operation the "implicitly
+     * new tuples" are now explicitly inserted this relation, and all of the new
+     * tuples in this relation are inserted into the old relation.
      */
-    void extend(const EquivalenceRelation<TupleType>& other) {
-        // nothing to extend if there's no new/original knowledge
-        if (other.size() == 0 || this->size() == 0) return;
+    void extendAndInsert(EquivalenceRelation<TupleType>& other) {
+        if (other.size() == 0 && this->size() == 0) return;
 
         std::set<value_type> repsCovered;
 
+        // This vector holds all of the elements of this equivalence relation
+        // that aren't yet in other, which get inserted after extending this
+        // relation by other. These operations are interleaved for maximum
+        // efficiency - either extend or inserting first would make the other
+        // operation unnecessarily slow.
+        std::vector<std::pair<value_type, value_type>> toInsert;
+        auto size = std::distance(this->sds.sparseToDenseMap.begin(), this->sds.sparseToDenseMap.end());
+        toInsert.reserve(size);
+
         // find all the disjoint sets that need to be added to this relation
         // that exist in other (and exist in this)
         {
```

Comment:
Have you experimented swapping the traversals to compute `repsCovered`? 
You could traverse over `others` and find the containment in `this`.

Reply:
Just in case you would like to further tune the performance...